File: Lessons/Lesson 01/BinaryGap.c
// you can write to stdout for debugging purposes, e.g.
// printf("this is a debug message\n");

int solution(int N) {
    int flag = 0;
    int counter = 0;
    int gap = 0;
    while (N) {
        int digit = (N & 1) ? 1 : 0;
        N >>= 1;
        if (!digit && flag) {
            counter++;
        } else if (digit && !flag) {
            flag = 1;
        } else {
            if (counter > gap) gap = counter;
            counter = 0;
        }
    }
    return gap;
}

File: Lessons/Lesson 01/BinaryGap.m
#import <Foundation/Foundation.h>
// you can also use other imports, for example:
// #import <SomeLibrary/SomeFile.h>

// you can write to stdout for debugging purposes, e.g.
// printf("this is a debug message\n");

int solution(int N) {
    NSMutableArray *binaryDigits = [NSMutableArray array];
    while (N > 0) {
        NSNumber *digit = (N & 1) ? @1 : @0;
        [binaryDigits insertObject:digit atIndex:0];
        N >>= 1;
    }
    NSInteger counter = 0;
    NSInteger gap = 0;
    for (NSNumber *digit in binaryDigits) {
        if ([digit isEqualToNumber:@0]) {
            counter++;
        } else {
            gap = MAX(gap, counter);
            counter = 0;
        }
    }
    return gap;
}
